beacon_node/network/src/sync/block_lookups/mod.rs

```diff
@@ -180,58 +276,133 @@ impl<T: BeaconChainTypes> BlockLookups<T> {
                         seen_timestamp,
                         BlockProcessType::SingleBlock { id },
                         cx,
-                    )
-                    .is_err()
-                {
-                    // Remove to avoid inconsistencies
-                    self.single_block_lookups.remove(&id);
+                    ) {
+                        Ok(()) => ShouldRemoveLookup::False,
+                        Err(()) => ShouldRemoveLookup::True,
+                    }
+                } else {
+                    ShouldRemoveLookup::False
                 }
             }
-            Ok(None) => {
-                // request finished correctly, it will be removed after the block is processed.
-            }
-            Err(error) => {
-                let msg: &str = error.into();
-                cx.report_peer(peer_id, PeerAction::LowToleranceError, msg);
-                // Remove the request, if it can be retried it will be added with a new id.
-                let mut req = request.remove();
-
-                debug!(self.log, "Single block lookup failed";
-                        "peer_id" => %peer_id, "error" => msg, "block_root" => %req.hash);
-                // try the request again if possible
-                if let Ok((peer_id, request)) = req.request_block() {
-                    if let Ok(id) = cx.single_block_lookup_request(peer_id, request) {
-                        self.single_block_lookups.insert(id, req);
+            Ok(None) => ShouldRemoveLookup::False,
+            Err(e) => handle_block_lookup_verify_error(
+                request_ref,
+                ResponseType::Block,
+                peer_id,
+                e,
+                cx,
+                &log,
+            ),
+        };
+
+        if matches!(should_remove, ShouldRemoveLookup::True) {
+            self.single_block_lookups
+                .retain(|req| req.id.block_request_id != Some(id));
+        }
+
+        metrics::set_gauge(
+            &metrics::SYNC_SINGLE_BLOCK_LOOKUPS,
+            self.single_block_lookups.len() as i64,
+        );
+    }
+
+    pub fn single_blob_lookup_response(
+        &mut self,
+        id: Id,
+        peer_id: PeerId,
+        blob: Option<Arc<BlobSidecar<T::EthSpec>>>,
+        seen_timestamp: Duration,
+        cx: &mut SyncNetworkContext<T>,
+    ) {
+        let stream_terminator = blob.is_none().into();
+
+        let log = self.log.clone();
+
+        let Some((has_pending_parent_requests, request_ref)) =
+            self.find_single_lookup_request(id, stream_terminator, ResponseType::Blob) else {
+            return;
+        };
+
+        let should_remove = match request_ref.verify_blob(blob) {
+            Ok(Some((block_root, blobs))) => {
+                if let Some(parent_components) = request_ref.unknown_parent_components.as_mut() {
+                    parent_components.add_unknown_parent_blobs(blobs);
+
+                    if !has_pending_parent_requests {
```

Comment:
I didn't understand why we only send for processing if there are no pending parent requests.

Reply:
Clear to me after speaking to sean offline. Will add some docs for this